### ISSUE TITLE
Worker should not panic on errors

### ISSUE THREAD

**Issue by dgoeke** [eramus/worker#1]
It's considered bad form for a Go library to panic on non-fatal errors, since it kills the entire application.  If it must panic, it should recover without passing the error up the stack.

As described in [Effective Go](https://golang.org/doc/effective_go.html#panic):

> This is only an example but real library functions should avoid panic. If the problem can be masked or worked around, it's always better to let things continue to run rather than taking down the whole program. One possible counterexample is during initialization: if the library truly cannot set itself up, it might be reasonable to panic, so to speak.

In this case, the library panics over bad JSON, connection issues, errors like "deadline soon," etc.  These should be returned to the caller in some way.

**Comment by eramus** [comment]
I agree with you that the library should not stop the application (and lets keep in mind that this is in no way production ready library), but the only places we are going to bail out are on the initial connection to beanstalkd or if we encounter a connection issue during operations. Ideally, this doesnt happen, but I decided to let the OS restart the application in these cases (upstart in my case).

The other cases you suggest are being recovered by the defer statement in `work()`. This will catch a panic when there is JSON marshaling error or when a response is sent back through beanstalkd. I made a conscious decision on my end to do this; simply because it just hasnt ever happened, and recovering and restarting the job runner seemed like a path of least resistance to ack the issue. It also has a side-effect of recovering panics from inside the actual job runner (third party to this library). Maybe this handling isnt 100% ideal though.

Im definitely open to suggestions for better handling these scenarios. This is a work in progress, but it is being used heavily for a side project of mine -- with decent success. Any suggestions/improvements/fixes are definitely welcome.
